### ISSUE TITLE
Is it possible to use ListGraphType<CustomType> inside of InputObjectGraphTypes?

### ISSUE THREAD

**Issue by sellingsolutions** [graphql-dotnet/graphql-dotnet#1831]
I've spen the last 3 hours trying to mimic https://gist.github.com/DanielRobinsonSoftware/9ef4aa2834f2c912b5a3d4ef63ba6188 as well as various other workarounds on stackoverflow. 

Afaik, the documentation doesn't mention how to use input arrays, I've so far had no issues with ListGraphType<StringGraphType> but I can't seem to understand how to use ListGraphType<MyOwnGraphType> inside of InputObjectGraphType.

This is all I want to do, to be able to pass in arrays with custom objects. What am I missing?
```
public class Input
{
        [JsonProperty("id")]
        public string Id { get; set; }
        
        [JsonProperty("rows")]
        public IEnumerable<Row> Rows { get; set; }
}
public class InputType : InputObjectGraphType<Input>
{
      public InputType() 
      {
          Field(o => o.Id, true, typeof(StringGraphType));
          Field(o => o.Rows, true, typeof(ListGraphType<RowType>));
      }
}
public class RowType : InputObjectGraphType<Row>
{
      public RowType() 
      {
          Field(o => o.Name, true, typeof(StringGraphType));
      }
}
```

**The exception that's being thrown**
```
"GraphQL.ExecutionError: Cannot convert value to AST: System.Collections.Generic.Dictionary`2[System.String,System.Object]\n ---> GraphQL.ExecutionError: Cannot convert value to AST: System.Collections.Generic.Dictionary`2[System.String,System.Object]\n   at GraphQL.GraphQLExtensions.AstFromValue(Object value, ISchema schema, IGraphType type)\n   at GraphQL.GraphQLExtensions.<>c__DisplayClass17_0.<AstFromValue>b__1(KeyValuePair`2 pair)\n   at System.Linq.Enumerable.SelectEnumerableIterator`2.ToList()\n   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)\n   at GraphQL.GraphQLExtensions.AstFromValue(Object value, ISchema schema, IGraphType type)\n   at GraphQL.GraphQLExtensions.AstFromValue(Object value, ISchema schema, IGraphType type)\n   at GraphQL.Execution.ExecutionHelper.GetVariableValue(Document document, ISchema schema, VariableDefinition variable, Object input)\n   at GraphQL.Execution.ExecutionHelper.GetVariableValues(Document document, ISchema schema, VariableDefinitions variableDefinitions, Inputs inputs)\n   at GraphQL.DocumentExecuter.BuildExecutionContext(ISchema schema, Object root, Document document, Operation operation, Inputs inputs, IDictionary`2 userContext, CancellationToken cancellationToken, Metrics metrics, IEnumerable`1 listeners, Boolean throwOnUnhandledException)\n   at GraphQL.DocumentExecuter.ExecuteAsync(ExecutionOptions options)\n   --- End of inner exception stack trace ---"]
```

**Comment by sungam3r** [comment]
What version do you use?

**Comment by sellingsolutions** [comment]
https://www.nuget.org/packages/GraphQL/3.0.0-preview-1175

**Comment by sungam3r** [comment]
Could you provide your query?

**Comment by Shane32** [comment]
This is a confirmed bug.  GraphQL literals work with the above sample, but variables do not.  See test class `Bug1831` that is pushed to branch `issues/1831`.  @sungam3r I've not looked at the cause yet; feel free to do so if you have the time.  One of us should fix it before 3.1 if possible.

**Comment by sungam3r** [comment]
This is an interesting bug. One bug hides another bug. In fact there is no `Bug1831InputGraphType`. `GraphType` suffix is removed by convention 😄 .

 See `KnownTypeNames` and `VariablesAreInputTypes` rules. Now `VariablesAreInputTypes` gives `Variable "arg" cannot be non-input type "Bug1831InputGraphType".` because `null` indeed is not input type.
![изображение](https://user-images.githubusercontent.com/21261007/93147475-fa02d280-f6f9-11ea-9377-ab143a7229d0.png)

If you comment `VariablesAreInputTypes` rule in `DocumentValidator.CoreRules` then you see _hidden_ error: `Unknown type Bug1831InputGraphType. Did you mean "Bug1831Input"?` validationResult has 2 errors but then `BuildExecutionContext` throw (inside `GetVariableValue`):


Howewer, now I see different message: `"Variable \u0027$arg\u0027 is invalid. Invalid input"`.

So... we can't build execution context because of bad variable type. Should we just return earlier? We can move 
```c#
  if (!validationResult.IsValid)
                {
                    return new ExecutionResult
                    {
                        Errors = validationResult.Errors,
                        Perf = metrics.Finish()
                    };
                }
```
just after 
```c#
 using (metrics.Subject("document", "Validating document"))
                {
                    validationResult = await _documentValidator.ValidateAsync(
                        options.Query,
                        options.Schema,
                        document,
                        options.ValidationRules,
                        options.UserContext,
                        options.Inputs);
                }
```

I also think that such an error could place in complexity analyzer as well.

**Comment by sungam3r** [comment]
So we will not call `AfterValidationAsync` on listeners. In principle, it's even good. It should be documented that this `AfterValidationAsync` is called only after **successful** validation.

**Comment by Shane32** [comment]
Ok so after I fixed the query then it does work.  The bug is that the validation error should be telling us that the type name is invalid.  Correct?

**Comment by sungam3r** [comment]
No. The bug is wrong order of operations. Validation result has two errors - that is right. But we lost this result because of unhandled exception when building execution context a bit later.

**Comment by Shane32** [comment]
Ok, right.

**Comment by Shane32** [comment]
I updated the branch to demonstrate and test the problem more precisely

**Comment by Shane32** [comment]
@sellingsolutions I was able to successfully use a a `ListGraphType<CustomType>` inside of an `InputObjectGraphType` as can be seen here:

https://github.com/graphql-dotnet/graphql-dotnet/blob/issues/1831/src/GraphQL.Tests/Bugs/Bug1831.cs

See `TestVariableObject()` and `TestLiteralObject()` which both pass successfully.  Perhaps you can look at the code and see what the difference is between that and your code is.

The only issue I had was that when submitting the query, the type of `Bug1831InputGraphType` is automatically changed to `Bug1831Input`.  You may be running into the same issue where you would need to use `Input` instead of `InputType` as the type argument within the query.  I'm not sure exactly without additional tests and seeing your exact query that is giving you problems.

The bug we have identified is that when writing a query with an invalid type name, the validation errors are not returned to the user.  We will try to get that fixed in 3.1 - but it is not the cause of your problem.

Please let me know if you have additional questions.

**Comment by Shane32** [comment]
@sungam3r Should I open a new issue or shall we track the bug within this issue?

**Comment by sungam3r** [comment]
Let's track it here.

**Comment by sungam3r** [comment]
> So we will not call AfterValidationAsync on listeners. In principle, it's even good. It should be documented that this AfterValidationAsync is called only after successful validation.

![изображение](https://user-images.githubusercontent.com/21261007/93456750-ce910b00-f8e6-11ea-8d06-446ed7eff10f.png)

What will we do next? Split `AfterValidationAsync` into two methods for successful and failed validation with different first params ? We just can't build execution context now for such an error. Or we can somehow fix the error with variable type so that we **can** build execution context (untyped variable value, object, special _unknown_ type)? Probably we can relax checks in `GetVariableValue` for such case when we know that we would not execute request later. We just need to create a execution context and pass it to listeners.

**Comment by Shane32** [comment]
Wouldn't it be better to pass it a `ValidationContext` instance (or maybe a `IValidationContext` instance) rather than the `IExecutionContext`?

**Comment by Shane32** [comment]
Just thinking out loud.... maybe not....

**Comment by sungam3r** [comment]
We need to define areas of responsibility.

**Comment by Shane32** [comment]
Yeah we already have `BeforeExecutionAsync(IExecutionContext context)` and there's basically no code between the call to `AfterValidationAsync` and that.  So changing `AfterValidationAsync` to have some type of validation context instance should work ....

**Comment by sungam3r** [comment]
One more breaking change 😉

**Comment by Shane32** [comment]
Right; that would need to go into 4.0.  Perhaps we should just patch 3.1 -- for example:
* if there are errors,
  * try to serialize the variables
  * if it fails, ignore the error and pass null for the variables prop into the context
  * then call AfterValidationAsync
    * if that removes the errors, then serialize the variables again which will throw an error.

another possibility: make the variables a lazy loading property of the IExecutionContext (only for AfterValidationAsync)

As for breaking changes, I don't have issues with breaking changes in 4.0 so long as we document them properly, and it's done for a good reason.

**Comment by sungam3r** [comment]
I do not mind breaking changes too.

**Comment by Shane32** [comment]
If we make breaking changes, I suggest we document them, even if sparsely, within the upgrade doc I prepped:

https://github.com/graphql-dotnet/graphql-dotnet/blob/develop/docs2/site/docs/guides/migration4.md

We can clean up the file before release, but it will be easy to track what's changed if we always add a bullet point when making a breaking change.

**Comment by sungam3r** [comment]
Good

**Comment by Shane32** [comment]
> Right; that would need to go into 4.0. Perhaps we should just patch 3.1 -- for example:
> 
> * if there are errors,
>   
>   * try to serialize the variables
>   * if it fails, ignore the error and pass null for the variables prop into the context
>   * then call AfterValidationAsync
>     
>     * if that removes the errors, then serialize the variables again which will throw an error.

I can make a PR along these lines.  It should not be a breaking change.  Any queries that worked before will continue to do so.  Any errors that are returned will be either return the same, or return a more accurate message.

Alternatively, we can do this:

> another possibility: make the variables a lazy loading property of the IExecutionContext (only for AfterValidationAsync)

This might actually be better, as it is guaranteed to execute the same for any implementation that has code within `AfterValidationAsync` and actually uses the `Variables` property -- but with the same problems if it is used.

Either way, in 4.0 we can make breaking changes to eliminate this workaround.

**Comment by sungam3r** [comment]
> Either way, in 4.0 we can make breaking changes to eliminate this workaround.

I don't know which of options is better, but I'm sure that the correct final change will be breaking.

**Comment by Shane32** [comment]
Lazy loading the `Variables` property is not possible as a non-breaking change since `Variables` is not declared `virtual` and `ExecutionContext` is not an interface.
